docs/source/app-dev/command-deduplication.rst

```diff
@@ -0,0 +1,352 @@
+Similarly, an :ref:`Exercise command <com.daml.ledger.api.v1.ExerciseCommand>` on a non-consuming choice or an :ref:`Exercise-By-Key command <com.daml.ledger.api.v1.ExercisebyKeyCommand>` may be executed multiple times if submitted multiple times.
+With command deduplication, applications can ensure such intended ledger changes are executed only once within the deduplication period, even if the application resubmits, say because it considers the earlier submissions to be lost or forgot during a crash that it had already submitted the command.
+
+As a typical example, the application wants to create exactly one contract instance of template ``T`` via the corresponding :ref:`Create command <com.daml.ledger.api.v1.CreateCommand>`.
+We assume that the ledger is correctly set up so that the Create command should be accepted, i.e., Daml packages are uploaded and vetted, the parties are allocated, and the application has a valid access token for the submission.
+
+
+Known processing time bounds
+============================
+
+For now, we assume that the application knows a bound ``B`` on the processing time and forward clock drifts in the Daml ledger w.r.t. the application’s clock.
+If processing measured across all retries takes longer than the bound ``B``, the ledger change may take effect several times.
+Under this assumption, the following strategy works for applications that use the :ref:`Command Service <command-service>` or the :ref:`Command Submission <command-submission-service>` and :ref:`Command Completion Service <command-completion-service>`.
+
+.. _dedup-bounded-step-command-id:
+
+#. Choose a fresh command ID for the ledger change and the ``actAs`` parties, which (together with the application ID) determine the change ID.
+   Remember the chosen command ID for the ledger change across application crashes.
+
+   .. note::
+      Make sure that you assign the same command ID to all command (re-)submissions of the same ledger change.
+      If you derive the command ID deterministically from the ledger change, you must only remember the ledger change, but not the command ID because you can always derive the command ID.
+
+      This is useful for the recovery procedure after an application crash/restart.
+      After a crash, the application in general cannot know whether it has submitted a set of commands before the crash.
+      If in doubt, resubmit the commands using the same command ID.
+      If the commands had been submitted before the crash, command deduplication on the ledger will reject the resubmissions.
+
+   .. _dedup-bounded-step-offset:
+
+#. When you use the :ref:`Command Completion Service <command-submission-service>`, obtain a recent offset on the completion stream ``OFF1``, say the :ref:`current ledger end <com.daml.ledger.api.v1.CommandCompletionService.CompletionEnd>`.
+
+   .. _dedup-bounded-step-submit:
+   
+#. Submit the command with the following parameters
```

Comment:
needs a colon at the end of the sentence to introduce the list of parametes

Reply:
Added.